File: ruckus-agent/insightfinder/insightfinder.go
package insightfinder

import (
	"bytes"
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"time"

	"github.com/carlmjohnson/requests"
	"github.com/google/go-querystring/query"
	config "github.com/insightfinder/ruckus-agent/configs"
	"github.com/insightfinder/ruckus-agent/pkg/models"
	"github.com/sirupsen/logrus"
)

const (
	METRIC_DATA_API     = "/api/v2/metric-data-receive"
	LOG_DATA_API        = "/api/v1/customprojectrawdata"
	PROJECT_ENDPOINT    = "/api/v1/check-and-add-custom-project"
	LOG_DATA_AGENT_TYPE = "Stream"
	CHUNK_SIZE          = 2 * 1024 * 1024 // 2MB
	MAX_PACKET_SIZE     = 10000000        // 10MB
	HTTP_RETRY_TIMES    = 3
	HTTP_RETRY_INTERVAL = 5 // seconds
)

func NewService(config config.InsightFinderConfig) *Service {
	// Create HTTP client with timeout and TLS settings
	tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}

	client := &http.Client{
		Transport: tr,
		Timeout:   180 * time.Second,
	}

	service := &Service{
		config:      config,
		httpClient:  client,
		ProjectName: config.ProjectName,
		SystemName:  config.SystemName,
		ProjectType: "Metric", // Default for Ruckus metrics
		Container:   config.IsContainer,
	}

	// Validate and set defaults
	service.Validate()

	return service
}

// Validate configuration and set defaults
func (s *Service) Validate() bool {
	logrus.Info("Validating InsightFinder configuration...")

	if s.config.ServerURL == "" {
		logrus.Error("ServerURL is required")
		return false
	}
	if s.config.UserName == "" {
		logrus.Error("UserName is required")
		return false
	}
	if s.config.LicenseKey == "" {
		logrus.Error("LicenseKey is required")
		return false
	}
	if s.config.ProjectName == "" {
		logrus.Error("ProjectName is required")
		return false
	}

	// Set defaults for missing fields
	if s.SystemName == "" {
		s.SystemName = s.ProjectName
		logrus.Warnf("SystemName not set, defaulting to ProjectName: %s", s.ProjectName)
	}

	if s.CloudType == "" {
		if s.config.CloudType != "" {
			s.CloudType = s.config.CloudType
		} else {
			s.CloudType = "OnPremise"
			logrus.Warn("CloudType not set, defaulting to OnPremise")
		}
	}

	if s.InstanceType == "" {
		if s.config.InstanceType != "" {
			s.InstanceType = s.config.InstanceType
		} else {
			s.InstanceType = "OnPremise"
			logrus.Warn("InstanceType not set, defaulting to OnPremise")
		}
	}

	// Set ProjectType from config or default
	if s.config.ProjectType != "" {
		s.ProjectType = s.config.ProjectType
	}

	// Set DataType based on ProjectType
	s.DataType = s.ProjectType

	// Set default sampling interval (from config or default to 5 minutes)
	if s.config.SamplingInterval > 0 {
		s.SamplingInterval = uint(s.config.SamplingInterval)
	} else {
		s.SamplingInterval = 5 * 60 // 5 minutes default
	}

	// Set InsightAgentType based on ProjectType and Container settings
	if s.ProjectType == "Metric" {
		if s.Container {
			s.InsightAgentType = "containerStreaming"
		} else {
			s.InsightAgentType = "RuckusAgent" // Custom agent type for Ruckus
		}
	} else {
		if s.Container {
			s.InsightAgentType = "ContainerCustom"
		} else {
			s.InsightAgentType = "RuckusAgent"
		}
	}

	logrus.Infof("InsightFinder configuration validated successfully:")
	logrus.Infof("  Project: %s", s.ProjectName)
	logrus.Infof("  System: %s", s.SystemName)
	logrus.Infof("  CloudType: %s", s.CloudType)
	logrus.Infof("  AgentType: %s", s.InsightAgentType)
	logrus.Infof("  SamplingInterval: %d seconds", s.SamplingInterval)

	return true
}

// Create project if it doesn't exist
func (s *Service) CreateProjectIfNotExist() bool {
	if !s.IsProjectExist() {
		return s.CreateProject()
	}
	return true
}

// Check if project exists
func (s *Service) IsProjectExist() bool {
	logrus.Infof("Checking if project '%s' exists in InsightFinder", s.ProjectName)

	form := url.Values{}
	form.Add("operation", "check")
	form.Add("userName", s.config.UserName)
	form.Add("licenseKey", s.config.LicenseKey)
	form.Add("projectName", s.ProjectName)
	form.Add("systemName", s.SystemName)

	url := fmt.Sprintf("%s%s", s.config.ServerURL, PROJECT_ENDPOINT)

	req, err := http.NewRequest("POST", url, bytes.NewBufferString(form.Encode()))
	if err != nil {
		logrus.Errorf("Failed to create project check request: %v", err)
		return false
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := s.httpClient.Do(req)
	if err != nil {
		logrus.Errorf("Failed to check project existence: %v", err)
		return false
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		logrus.Errorf("Project check failed with status: %d", resp.StatusCode)
		return false
	}

	var checkResponse CheckAndAddCustomProjectResponse
	if err := json.NewDecoder(resp.Body).Decode(&checkResponse); err != nil {
		logrus.Errorf("Failed to decode project check response: %v", err)
		return false
	}

	if checkResponse.IsSuccess {
		if checkResponse.IsProjectExist {
			logrus.Infof("Project '%s' exists in InsightFinder", s.ProjectName)
			return true
		}
		logrus.Infof("Project '%s' does not exist in InsightFinder", s.ProjectName)
		return false
	}

	logrus.Errorf("Project check failed: %s", checkResponse.Message)
	return false
}

// Create new project
func (s *Service) CreateProject() bool {
	logrus.Infof("Creating project '%s' in InsightFinder", s.ProjectName)

	request := CheckAndAddCustomProjectRequest{
		Operation:        "create",
		UserName:         s.config.UserName,
		LicenseKey:       s.config.LicenseKey,
		ProjectName:      s.ProjectName,
		SystemName:       s.SystemName,
		InstanceType:     s.InstanceType,
		ProjectCloudType: s.CloudType,
		DataType:         s.DataType,
		InsightAgentType: s.InsightAgentType,
		SamplingInterval: int(s.SamplingInterval),
	}
	requestForm, err := query.Values(request)
	if err != nil {
		logrus.Errorf("Error building request form to create project: %v", err)
		return false
	}

	var resultStr string
	err = requests.URL(s.config.ServerURL).Path(PROJECT_ENDPOINT).Header("agent-type", "Stream").BodyJSON(request).Params(requestForm).ToString(&resultStr).Post().Fetch(context.Background())
	if err != nil {
		logrus.Errorf("Failed to create project: %v", err)
		return false
	}

	logrus.Infof("Project '%s' created successfully in InsightFinder", s.ProjectName)
	return true
}

// Send metrics in bulk - most efficient for large datasets
func (s *Service) SendMetricsBulk(metrics []models.MetricData) error {
	if len(metrics) == 0 {
		return nil
	}

	logrus.Infof("Sending %d metrics in bulk to InsightFinder using v2 API", len(metrics))

	// Group metrics by instance and timestamp
	instanceDataMap := make(map[string]InstanceData)
	var minTimestamp, maxTimestamp int64

	for i, metric := range metrics {
		instanceName := metric.InstanceName
		timestamp := metric.Timestamp

		// Convert timestamp to epoch milliseconds if it's in seconds
		if timestamp < 1e12 {
			timestamp = timestamp * 1000
		}

		// Track min/max timestamps
		if i == 0 {
			minTimestamp = timestamp
			maxTimestamp = timestamp
		} else {
			if timestamp < minTimestamp {
				minTimestamp = timestamp
			}
			if timestamp > maxTimestamp {
				maxTimestamp = timestamp
			}
		}

		// Get or create instance data
		instanceData, exists := instanceDataMap[instanceName]
		if !exists {
			instanceData = InstanceData{
				InstanceName:       instanceName,
				ComponentName:      instanceName,
				DataInTimestampMap: make(map[int64]DataInTimestamp),
			}
		}

		// Create metric data points
		metricDataPoints := make([]MetricDataPoint, 0)
		dataInTimestamp := DataInTimestamp{
			TimeStamp:        timestamp,
			MetricDataPoints: &metricDataPoints,
		}

		// Add all metric data points for this AP
		for metricName, value := range metric.Data {
			if floatValue, ok := convertToFloat64(value); ok {
				*dataInTimestamp.MetricDataPoints = append(*dataInTimestamp.MetricDataPoints, MetricDataPoint{
					MetricName: metricName,
					Value:      floatValue,
				})
			} else {
				logrus.Debugf("Skipping non-numeric metric %s: %v (type: %T)", metricName, value, value)
			}
		}

		instanceData.Zone = metric.Zone
		instanceData.DataInTimestampMap[timestamp] = dataInTimestamp
		instanceDataMap[instanceName] = instanceData
	}

	// Create payload using the v2 API format
	payload := IFMetricPostRequestPayload{
		LicenseKey: s.config.LicenseKey,
		UserName:   s.config.UserName,
		Data: MetricDataReceivePayload{
			ProjectName:      s.ProjectName,
			UserName:         s.config.UserName,
			SystemName:       s.SystemName,
			InstanceDataMap:  instanceDataMap,
			MinTimestamp:     minTimestamp,
			MaxTimestamp:     maxTimestamp,
			InsightAgentType: s.InsightAgentType,
			SamplingInterval: strconv.Itoa(int(s.SamplingInterval)),
			CloudType:        s.CloudType,
		},
	}

	return s.sendMetricPayloadV2(payload)
}

// Send metrics in streaming chunks - memory efficient
func (s *Service) SendMetricsStreaming(metrics []models.MetricData, chunkSize int) error {
	if len(metrics) == 0 {
		return nil
	}

	logrus.Infof("Sending %d metrics in streaming chunks of %d", len(metrics), chunkSize)

	for i := 0; i < len(metrics); i += chunkSize {
		end := i + chunkSize
		if end > len(metrics) {
			end = len(metrics)
		}

		chunk := metrics[i:end]
		logrus.Debugf("Sending chunk %d/%d (%d metrics)", (i/chunkSize)+1, (len(metrics)+chunkSize-1)/chunkSize, len(chunk))

		if err := s.SendMetricsBulk(chunk); err != nil {
			return fmt.Errorf("failed to send chunk %d: %v", (i/chunkSize)+1, err)
		}

		// Small delay between chunks
		if end < len(metrics) {
			time.Sleep(100 * time.Millisecond)
		}
	}

	return nil
}

// Send individual metrics - for backward compatibility
func (s *Service) SendMetrics(metrics []models.MetricData) error {
	return s.SendMetricsBulk(metrics)
}

// Send metric payload to InsightFinder v2 API
func (s *Service) sendMetricPayloadV2(payload IFMetricPostRequestPayload) error {
	url := fmt.Sprintf("%s%s", s.config.ServerURL, METRIC_DATA_API)

	jsonPayload, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("failed to marshal metrics payload: %v", err)
	}

	if len(jsonPayload) > MAX_PACKET_SIZE {
		return fmt.Errorf("payload size (%d bytes) exceeds maximum allowed (%d bytes)", len(jsonPayload), MAX_PACKET_SIZE)
	}

	logrus.Debugf("Sending %d bytes to InsightFinder v2 API", len(jsonPayload))

	// Send with retry logic
	var lastErr error
	for attempt := 1; attempt <= HTTP_RETRY_TIMES; attempt++ {
		req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonPayload))
		if err != nil {
			return fmt.Errorf("failed to create request: %v", err)
		}

		req.Header.Set("Content-Type", "application/json")

		resp, err := s.httpClient.Do(req)
		if err != nil {
			lastErr = err
			logrus.Warnf("Attempt %d/%d failed: %v", attempt, HTTP_RETRY_TIMES, err)
			if attempt < HTTP_RETRY_TIMES {
				time.Sleep(time.Duration(HTTP_RETRY_INTERVAL) * time.Second)
				continue
			}
			return fmt.Errorf("failed to send metrics after %d attempts: %v", HTTP_RETRY_TIMES, err)
		}
		defer resp.Body.Close()

		if resp.StatusCode == http.StatusOK {
			logrus.Infof("Successfully sent metrics to InsightFinder v2 API (instances: %d)", len(payload.Data.InstanceDataMap))
			return nil
		} else {
			bodyBytes := make([]byte, 1024)
			n, _ := resp.Body.Read(bodyBytes)
			lastErr = fmt.Errorf("request failed with status: %d, body: %s", resp.StatusCode, string(bodyBytes[:n]))
			logrus.Warnf("Attempt %d/%d failed with status %d", attempt, HTTP_RETRY_TIMES, resp.StatusCode)
			if attempt < HTTP_RETRY_TIMES {
				time.Sleep(time.Duration(HTTP_RETRY_INTERVAL) * time.Second)
				continue
			}
		}
	}

	return fmt.Errorf("failed to send metrics after %d attempts: %v", HTTP_RETRY_TIMES, lastErr)
}

// Get configuration
func (s *Service) GetConfig() config.InsightFinderConfig {
	return s.config
}

File: telrad-agent/insightfinder.py
from __future__ import annotations

import dataclasses
import json
import logging
import urllib.parse
from typing import Any

import requests
import urllib3
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

# Suppress SSL warnings
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

# API endpoint constants
API_METRIC_DATA_RECEIVE = "api/v2/metric-data-receive"
API_CHECK_ADD_PROJECT = "api/v1/check-and-add-custom-project"
API_CUSTOM_PROJECT_RAW_DATA = "api/v1/customprojectrawdata"
LOG_CHUNK_SIZE = 5000

logger = logging.getLogger(__name__)


def _build_session() -> requests.Session:
    """Create a requests session with retry logic for transient errors."""
    s = requests.Session()
    retry = Retry(
        total=3,
        connect=3,
        read=3,
        status=3,
        backoff_factor=1.0,
        backoff_jitter=0.3,
        status_forcelist=(429, 500, 502, 503, 504),
        allowed_methods=("POST",),
        respect_retry_after_header=True,
        raise_on_status=False,
    )
    adapter = HTTPAdapter(max_retries=retry)
    s.mount("http://", adapter)
    s.mount("https://", adapter)
    return s


@dataclasses.dataclass
class Config:
    url: str
    user_name: str
    license_key: str
    project_name: str
    agent_type: str
    instance_type: str = "PrivateCloud"
    chunk_size: int = 100000
    create_project: bool = False
    system_name: str | None = None
    data_type: str | None = None
    insight_agent_type: str | None = None
    sampling_interval: int = 600


class InsightFinder:
    """Client for interacting with InsightFinder API."""

    def __init__(self, config: Config) -> None:
        self.config = config
        self.session = _build_session()

        if self.config.create_project and not self.project_existed():
            self.create_custom_project()

    def __enter__(self) -> "InsightFinder":
        return self

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        self.close()

    def close(self) -> None:
        if self.session:
            self.session.close()

    def _build_base_payload(self) -> dict[str, str]:
        return {
            "userName": self.config.user_name,
            "licenseKey": self.config.license_key,
            "projectName": self.config.project_name,
        }

    def send_metric(self, data: dict[str, Any]) -> None:
        """Send metric data to InsightFinder v2 API.

        Args:
            data: idm (instance data map) structure
        """
        if not data:
            logger.warning("No data to send")
            return

        url = urllib.parse.urljoin(self.config.url, API_METRIC_DATA_RECEIVE)
        payload = {
            "userName": self.config.user_name,
            "licenseKey": self.config.license_key,
            "data": {
                "projectName": self.config.project_name,
                "userName": self.config.user_name,
                "iat": self.config.insight_agent_type or "Custom",
                "ct": self.config.instance_type,
                "idm": data,
            },
        }

        num_instances = len(data)
        logger.info(f"Sending data for {num_instances} instance(s)")
        try:
            self._request(url, payload)
            logger.info(f"Successfully sent data for {num_instances} instance(s)")
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to send data: {e}")
            raise

    def send_log(self, logs: list[dict[str, Any]]) -> None:
        """Send log records to InsightFinder v1 customprojectrawdata endpoint.

        Each record: {"timestamp": int_ms, "tag": str, "data": str, "componentName"?: str}
        Sends in chunks of 5000 records per the API guidance.

        Args:
            logs: List of log records
        """
        if not logs:
            logger.warning("No log data to send")
            return

        url = urllib.parse.urljoin(self.config.url, API_CUSTOM_PROJECT_RAW_DATA)
        total_chunks = (len(logs) + LOG_CHUNK_SIZE - 1) // LOG_CHUNK_SIZE
        for chunk_num, i in enumerate(range(0, len(logs), LOG_CHUNK_SIZE), start=1):
            chunk = logs[i : i + LOG_CHUNK_SIZE]
            payload = {
                "userName": self.config.user_name,
                "licenseKey": self.config.license_key,
                "projectName": self.config.project_name,
                "agentType": self.config.agent_type,
                "metricData": json.dumps(chunk),
            }
            logger.debug(f"Sending log chunk {chunk_num}/{total_chunks} ({len(chunk)} records)")
            try:
                self._request_form(url, payload)
                logger.info(f"Log chunk {chunk_num}/{total_chunks} sent ({len(chunk)} records)")
            except requests.exceptions.RequestException as e:
                logger.error(f"Failed to send log chunk {chunk_num}: {e}")
                raise

    def project_existed(self) -> bool:
        """Check if project exists in InsightFinder."""
        url = urllib.parse.urljoin(self.config.url, API_CHECK_ADD_PROJECT)
        payload = self._build_base_payload()
        payload["operation"] = "check"

        try:
            response = self._request_form(url, payload)
            content = response.json()
            logger.debug(f"Project check response: {content}")
            exists = content.get("isProjectExist", False)
            if exists:
                logger.info(f"Project '{self.config.project_name}' exists")
            else:
                logger.info(f"Project '{self.config.project_name}' does not exist")
            return exists
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to check project '{self.config.project_name}': {e}")
            return False
        except (ValueError, KeyError) as e:
            logger.error(f"Invalid response when checking project: {e}")
            return False

    def create_custom_project(self) -> bool:
        """Create a new project in InsightFinder."""
        url = urllib.parse.urljoin(self.config.url, API_CHECK_ADD_PROJECT)
        payload = self._build_base_payload()
        payload["operation"] = "create"
        payload["systemName"] = self.config.system_name or ""
        payload["dataType"] = self.config.data_type or ""
        payload["instanceType"] = self.config.instance_type
        payload["insightAgentType"] = self.config.insight_agent_type or ""
        payload["projectCloudType"] = "PrivateCloud"
        payload["samplingInterval"] = str(self.config.sampling_interval)

        try:
            response = self._request_form(url, payload)
            content = response.json()
            logger.debug(f"Project creation response: {content}")
            success = content.get("success", False)
            if success:
                logger.info(f"Project '{self.config.project_name}' created successfully")
            else:
                logger.warning(f"Project '{self.config.project_name}' creation failed")
            return success
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to create project '{self.config.project_name}': {e}")
            return False
        except (ValueError, KeyError) as e:
            logger.error(f"Invalid response when creating project: {e}")
            return False

    def _request(self, url: str, payload: dict[str, Any]) -> requests.Response:
        """POST JSON payload (v2 API)."""
        logger.debug(f"Sending request to {url}")
        headers = {"Content-Type": "application/json"}
        response = self.session.post(url, json=payload, headers=headers, verify=False, timeout=30)
        response.raise_for_status()
        return response

    def _request_form(self, url: str, payload: dict[str, str]) -> requests.Response:
        """POST form-encoded payload (v1 API)."""
        logger.debug(f"Sending request to {url}")
        headers = {"Content-Type": "application/x-www-form-urlencoded"}
        response = self.session.post(url, data=payload, headers=headers, verify=False, timeout=30)
        response.raise_for_status()
        return response
